src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+package mate.academy.service.impl;
+
+import lombok.RequiredArgsConstructor;
+import mate.academy.dto.cartitem.CartItemRequestDto;
+import mate.academy.dto.cartitem.UpdateCartItemDto;
+import mate.academy.dto.shoppingcart.ShoppingCartDto;
+import mate.academy.exception.EntityNotFoundException;
+import mate.academy.mapper.CartItemMapper;
+import mate.academy.mapper.ShoppingCartMapper;
+import mate.academy.model.Book;
+import mate.academy.model.CartItem;
+import mate.academy.model.ShoppingCart;
+import mate.academy.model.User;
+import mate.academy.repository.BookRepository;
+import mate.academy.repository.CartItemRepository;
+import mate.academy.repository.ShoppingCartRepository;
+import mate.academy.service.ShoppingCartService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final CartItemRepository itemRepository;
+    private final ShoppingCartRepository repository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper cartMapper;
+    private final CartItemMapper itemMapper;
+
+    @Override
+    public void registerShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        repository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto getUsersCart(Long userId) {
+        return cartMapper.toDto(repository.findByUserId(userId));
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto itemDto) {
+        Book book = bookRepository.findById(itemDto.bookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Book with id: %d is not found", itemDto.bookId())
+                ));
+        ShoppingCart cart = repository.findByUserId(userId);
+        CartItem cartItem = itemRepository.findByShoppingCartIdAndBookId(cart.getId(), book.getId())
+                .map(item -> {
+                    item.setQuantity(item.getQuantity() + itemDto.quantity());
+                    return item;
+                }).orElseGet(() -> createCartItem(itemDto, book, cart));
```

Comment:
do you really need to do it? your logic looks like you have a separate method for adding books to the cart and updating the quantity

Reply:
I just thought that if I click the addToCart button first - I add the book to the cart, if I click it for the second time I don't get any exception , the number of books in the cart will be updated . The requirements are very unclear and there is no product owner to consult (therefore I have to answer some questions myself =) ) That is, if the book is not in the cart - it will be added, if there is already such a book in the cart, then a new quantity will be added to the quantity